MaintenanceJobDesignDoc.md

```diff
@@ -0,0 +1,140 @@
+Maintenance Job Design Document
+===============================
+
+The (shared) object cache is growing to be large. This includes:
+
+* Number of loose objects.
+* Number of pack-files.
+* Size of pack-files.
+
+There are some things we can do to improve this situation, with varying levels
+of difficulty and benefit. Here are the four items I believe will have a strong
+impact on our data size and performance:
+
+I. Delete loose objects that exist in pack-files.
+II. Place loose objects into pack-files.
+III. Repack some prefetch packs into larger packs.
+IV. Delete packs with no objects referenced by the multi-pack-index.
+
+These are numbered by order of implementation, but they work together in an
+interesting way. If we place loose objects into pack-files (I), then the next
+time we delete loose objects (II) we will delete those objects that we just put
+into a pack-file. If we repack some prefetch packs into larger packs (III), the
+multi-pack-index will use the "newer" pack for all objects in those packs, and
+we can delete them (IV).
+
+In order to perform this maintenance without disrupting users, we can actually
+do these interactions in two independent jobs, with their sub-actions in a
+particular order:
+
+A. Loose Object Maintenance
+  1. Delete loose objects that exist in pack-files.
+  2. Place loose objects into pack-files.
+B. Prefetch Pack Maintenance
+  1. Delete packs with no objects referenced by the multi-pack-index.
+  2. Repack some prefetch packs into larger packs.
+
+Care must be taken to ensure that Git processes that initialized their object
+store before the maintenance   still succeed. Thus, we should never delete
+anything that those processes could be reading from. This is why we delete loose
+objects only if they are referenced by pack-files. This is why we delete
+prefetch packs only if they have no objects referenced by the multi-pack-index.
+Since a second run of these jobs could delete objects that were repacked in the
+previous run, we need to space our job runs such that no Git processes span both
+runs.
+
+Loose Object Job
+--------------
+
+The loose object job can save users who have many loose object downloads. These
+loose objects are populated by several user activities: running `git fetch` will
+get loose commits and trees for the ref tips that are not covered by prefetch
+packs, hydrating a file will trigger a loose blob, and many other. Older
+enlistments have many loose commits and trees due to the behavior of `git fetch`
+before the `--no-show-forced-updates` option.
+
+### Delete loose objects that appear in a pack-file
+
+By running `git prune-packed -q` with the environment variable
+`GIT_OBJECT_DIRECTORY` set to the object cache, Git will delete the loose
+objects that exist in a pack-file. This will use the multi-pack-index to test
+containment, so is only bound by the file-deletion operations.
+
+### Pack loose objects that do not appear in a pack-file

```

Comment:
Will this only be for loose object in the cache or loose object in the repo as well?

Reply:
Currently, I'm focused on the object cache, as that is where the majority of objects are. We could expand this to include objects in the .git folder of each enlistment (objects created by the user).

Reply:
I would make it an explicit design goal that we only clean up objects in our own cache. I think we should never mess with user-created objects.

Reply:
While I agree that this design should focus on the cache, I think that eventually we will need to help users clean up their .git folder. While they have many fewer local loose objects, as we have users keep enlistments for longer and longer, we will see performance degrade due to that, too. Normally, `git gc --auto` would kick in and do it for them, but we've disabled this command.

Reply:
Yes we should make it _possible_ for a user to clean up their git objects, but I don't think it's something we should do as a proactive cleanup. We own the cache folder, and we should clean that up, and we should treat the repo's .git folder as the user's own data and not mess with it without an explicit gesture. Two ways to handle that particular issue down the road would be to re-enable `git gc` to work in VFS for Git repos, or we can add a new verb of our own to do repo-local cleanup. Either way, outside the scope of a cache maintenance feature.

Reply:
You're right. It shouldn't be in the background. Something to consider as an extension to the manual way to trigger maintenance. Perhaps `gvfs maintenance --local` or something similar. I'll drop it for now.